File: sample-java/src/test/java/sample/banana/java/FifthTest.java
package sample.banana.java;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Field;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.Mockito.mock;

final class FifthTest {
  private First.Second.Third.Fourth.Fifth subject;

  @BeforeEach
  void setUp() {
    subject = new First.Second().new Third<>().new Fourth().new Fifth<String>();
  }

  @Test
  void setAField() throws NoSuchFieldException, IllegalAccessException {
    @SuppressWarnings("unchecked") final Set<String> expected = mock(Set.class);

    FirstSecondThirdFourthFifthTestAccessors.setYetAnotherField(subject, expected);

    final Field field = subject.getClass().getDeclaredField("yetAnotherField");
    final boolean wasAccessible = field.isAccessible();
    field.setAccessible(true);
    @SuppressWarnings("unchecked") final Set<String> actual = (Set<String>) field.get(subject);
    assertSame(expected, actual);
    field.setAccessible(wasAccessible);
  }

  @Test
  void getAField() throws NoSuchFieldException, IllegalAccessException {
    @SuppressWarnings("unchecked") final Set<String> expected = mock(Set.class);
    final Field field = subject.getClass().getDeclaredField("yetAnotherField");
    final boolean wasAccessible = field.isAccessible();
    field.setAccessible(true);
    field.set(subject, expected);
    field.setAccessible(wasAccessible);

    final Set<String> actual = FirstSecondThirdFourthFifthTestAccessors.getYetAnotherField(subject);

    assertSame(expected, actual);
  }
}

File: sample-kotlin/src/test/kotlin/sample/banana/kotlin/FifthTest.kt
package sample.banana.kotlin

import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import sample.banana.kotlin.FirstSecondThirdFourthFifthTestAccessors.yetAnotherField

internal class FifthTest {
	private lateinit var subject: First.Second.Third<Any>.Fourth.Fifth<String>

	@BeforeEach
	fun setUp() {
		subject = First.Second().Third<Any>().Fourth().Fifth()
	}

	@Test
	fun setAField() {
		@Suppress("UNCHECKED_CAST") val expected = mock(Set::class.java) as Set<String>

		subject.yetAnotherField(expected)

		var actual: Set<String>
		subject::class.java.getDeclaredField("yetAnotherField").apply {
			val wasAccessible = isAccessible
			isAccessible = true
			@Suppress("UNCHECKED_CAST")
			actual = this[subject] as Set<String>
			isAccessible = wasAccessible
		}
		assertSame(expected, actual)
	}

	@Test
	fun getAField() {
		@Suppress("UNCHECKED_CAST") val expected = mock(Set::class.java) as Set<String>
		subject.javaClass.getDeclaredField("yetAnotherField").apply {
			val wasAccessible = isAccessible
			isAccessible = true
			set(subject, expected)
			isAccessible = wasAccessible
		}

		val actual: Set<String> = subject.yetAnotherField()

		assertSame(expected, actual)
	}
}
